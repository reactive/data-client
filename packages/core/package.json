{
  "name": "@data-client/core",
  "version": "0.14.18",
  "description": "Async State Management without the Management. REST, GraphQL, SSE, Websockets, Fetch",
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "next": [
        "lib/next/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "next": [
        "ts3.4/next/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./next": {
      "types": "./lib/next/index.d.ts",
      "require": "./dist/next.js",
      "default": "./lib/next/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md",
    "./typescript.svg",
    "./data_client_logo_and_text.svg"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:js:node": "BROWSERSLIST_ENV=node12 yarn g:rollup",
    "build:js:browser": "BROWSERSLIST_ENV=legacy yarn g:rollup",
    "build:bundle": "yarn g:runs build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "yarn g:clean ./index.d.ts",
    "build:legacy-types": "yarn g:downtypes lib ts3.4",
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "dev": "run build:lib -w",
    "prepare": "run build:lib",
    "prepack": "run prepare && run build:bundle && run build:legacy:lib"
  },
  "keywords": [
    "data",
    "cache",
    "normalizr",
    "normalization",
    "flux",
    "fetch",
    "hook",
    "networking",
    "async",
    "typescript",
    "async",
    "data fetching",
    "data cache",
    "reactive",
    "state management",
    "api client",
    "api",
    "normalized cache",
    "swr",
    "query",
    "ios",
    "android",
    "web",
    "middleware",
    "websocket",
    "SSE",
    "GraphQL",
    "REST",
    "RPC",
    "declarative"
  ],
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "license": "Apache-2.0",
  "homepage": "https://dataclient.io",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "@data-client/normalizr": "^0.14.16",
    "core-js-pure": "^3.40.0",
    "flux-standard-action": "^2.1.1"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/endpoint": "workspace:*",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "rollup-plugins": "workspace:*"
  }
}
