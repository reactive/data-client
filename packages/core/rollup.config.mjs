import {
  babel,
  commonjs,
  filesize,
  json,
  resolve,
  replace,
  terser,
  typeConfig,
  typeConfigNext,
  onwarn,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies)
  //.concat(Object.keys(pkg.peerDependencies))
  .filter(dep => !['@babel/runtime'].includes(dep));

const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node', '.jsx'];
const nativeExtensions = ['.native.ts', ...extensions];
process.env.NODE_ENV = 'production';

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];
if (process.env.BROWSERSLIST_ENV !== 'node12') {
  // browser-friendly UMD build
  configs.push({
    input: 'src/index.ts',
    output: [{ file: pkg.unpkg, format: 'umd', name: 'RDC.Core' }],
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        rootMode: 'upward',
        babelHelpers: 'runtime',
        caller: { polyfillMethod: false },
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
        preventAssignment: true,
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      json(),
      terser({}),
      filesize({ showBrotliSize: true }),
    ],
  });
  configs.push(typeConfig);
  configs.push(typeConfigNext);
} else {
  // node-friendly commonjs build
  [
    { input: 'src/index.ts', output: pkg.main },
    { input: 'src/next/index.ts', output: 'dist/next.js' },
  ].forEach(({ input, output }) => {
    configs.push({
      input,
      external: isExternal,
      output: [{ file: output, format: 'cjs' }],
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '**/__tests__/**', '**/*.d.ts'],
          extensions: nativeExtensions,
          rootMode: 'upward',
          babelHelpers: 'runtime',
        }),
        replace({ 'process.env.CJS': 'true', preventAssignment: true }),
        resolve({ extensions: nativeExtensions }),
        commonjs({ extensions: nativeExtensions }),
      ],
    });
  });
}
export default configs;
