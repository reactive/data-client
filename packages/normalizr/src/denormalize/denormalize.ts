import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
