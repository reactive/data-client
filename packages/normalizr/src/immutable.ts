export { DelegateImmutable } from './memo/Delegate.immutable.js';
