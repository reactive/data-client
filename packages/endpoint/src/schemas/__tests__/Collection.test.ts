// eslint-env jest
import { initialState } from '@data-client/core';
import { normalize, denormalize, MemoCache } from '@data-client/normalizr';
import { ArticleResource, IDEntity } from '__tests__/new';
import { Record } from 'immutable';

import { SimpleMemoCache } from './denormalize';
import { PolymorphicInterface } from '../..';
import { schema } from '../..';
import PolymorphicSchema from '../Polymorphic';

let dateSpy: jest.SpyInstance;
beforeAll(() => {
  dateSpy = jest
    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

class Todo extends IDEntity {
  userId = 0;
  title = '';
  completed = false;

  static key = 'Todo';
}

class User extends IDEntity {
  name = '';
  username = '';
  email = '';
  todos: Todo[] = [];

  static key = 'User';
  static schema = {
    todos: new schema.Collection(new schema.Array(Todo), {
      nestKey: (parent, key) => ({
        userId: parent.id,
      }),
    }),
  };
}

const userTodos = new schema.Collection(new schema.Array(Todo), {
  argsKey: ({ userId }: { userId: string }) => ({
    userId,
  }),
});

test('key works with custom schema', () => {
  class CustomArray extends PolymorphicSchema {
    declare schema: any;
    normalize(
      input: any,
      parent: any,
      key: any,
      args: any[],
      visit: any,
      snapshot: any,
    ): any {
      return input.map((value, index) =>
        this.normalizeValue(value, parent, key, args, visit),
      );
    }

    denormalize(
      input: any,
      args: any[],
      unvisit: (schema: any, input: any) => any,
    ) {
      return input.map ?
          input.map((entityOrId: any) =>
            this.denormalizeValue(entityOrId, unvisit),
          )
        : input;
    }

    queryKey(args: unknown, unvisit: unknown, delegate: unknown): any {
      return undefined;
    }

    _normalizeNullable(): any {}

    _denormalizeNullable(): any {}
  }

  const collection = new schema.Collection(new CustomArray(Todo));
  expect(collection.key).toBe('(Todo)');
});

describe(`${schema.Collection.name} normalization`, () => {
  let warnSpy: jest.SpyInstance;
  afterEach(() => {
    warnSpy.mockRestore();
  });
  beforeEach(() =>
    (warnSpy = jest.spyOn(console, 'warn')).mockImplementation(() => {}),
  );

  test('should throw a custom error if data loads with string unexpected value', () => {
    function normalizeBad() {
      normalize(userTodos, 'abc');
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('should throw a custom error if data loads with string unexpected value', () => {
    function normalizeBad() {
      normalize(userTodos.push, null);
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('should throw a custom error if data loads with args of unexpected value', () => {
    function normalizeBad() {
      userTodos.normalize(
        [],
        undefined as any,
        '',
        [],
        () => undefined,
        {} as any,
      );
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('normalizes nested collections', () => {
    const state = normalize(User, {
      id: '1',
      username: 'bob',
      todos: [{ id: '5', title: 'finish collections' }],
    });
    expect(state).toMatchSnapshot();
    //const a: string | undefined = state.result;
    // @ts-expect-error
    const b: Record<any, any> | undefined = state.result;
  });

  test('normalizes top level collections', () => {
    const state = normalize(
      userTodos,
      [{ id: '5', title: 'finish collections' }],
      [{ userId: '1' }],
    );
    expect(state).toMatchSnapshot();
    //const a: string[] | undefined = state.result;
    // @ts-expect-error
    const b: Record<any, any> | undefined = state.result;
  });

  test('normalizes top level collections (no args)', () => {
    const state = normalize(
      new schema.Collection(new schema.Array(Todo)),
      [{ id: '5', title: 'finish collections' }],
      [{ userId: '1' }],
    );
    expect(state).toMatchSnapshot();
    //const a: string[] | undefined = state.result;
    // @ts-expect-error
    const b: Record<any, any> | undefined = state.result;
  });

  test('normalizes already processed entities', () => {
    const state = normalize(User, {
      id: '1',
      username: 'bob',
      todos: ['5', '6'],
    });
    expect(state).toMatchSnapshot();
  });

  describe('polymorphism', () => {
    class User extends IDEntity {
      type = 'users';
    }
    class Group extends IDEntity {
      type = 'groups';
    }
    const collection = new schema.Collection(
      new schema.Array(
        {
          users: User,
          groups: Group,
        },
        'type',
      ),
    );
    const collectionUnion = new schema.Collection([
      new schema.Union(
        {
          users: User,
          groups: Group,
        },
        'type',
      ),
    ]);

    test('generates polymorphic key', () => {
      expect(collection.key).toBe('[User;Group]');
      expect(collectionUnion.key).toBe('[User;Group]');
    });

    test('works with polymorphic members', () => {
      const { entities, result } = normalize(
        collection,
        [
          { id: '1', type: 'users' },
          { id: '2', type: 'groups' },
        ],
        [{ fakeFilter: false }],
      );
      expect(result).toMatchSnapshot();
      expect(entities).toMatchSnapshot();
    });
    test('works with Union members', () => {
      const { entities, result } = normalize(
        collectionUnion,
        [
          { id: '1', type: 'users' },
          { id: '2', type: 'groups' },
        ],
        [{ fakeFilter: false }],
      );
      expect(result).toMatchSnapshot();
      expect(entities).toMatchSnapshot();
    });
  });

  test('normalizes push onto the end', () => {
    const init = {
      entities: {
        [User.schema.todos.key]: {
          '{"userId":"1"}': ['5'],
        },
        Todo: {
          '5': {
            id: '5',
            title: 'finish collections',
          },
        },
        User: {
          '1': {
            id: '1',
            todos: '{"userId":"1"}',
            username: 'bob',
          },
        },
      },
      entityMeta: {
        [User.schema.todos.key]: {
          '{"userId":"1"}': {
            date: 1557831718135,
            expiresAt: Infinity,
            fetchedAt: 0,
          },
        },
        Todo: {
          '5': {
            date: 1557831718135,
            expiresAt: Infinity,
            fetchedAt: 0,
          },
        },
        User: {
          '1': {
            date: 1557831718135,
            expiresAt: Infinity,
            fetchedAt: 0,
          },
        },
      },
      indexes: {},
    };
    const state = normalize(
      User.schema.todos.push,
      [{ id: '10', title: 'create new items' }],
      [{ userId: '1' }],
      init,
    );
    expect(state).toMatchSnapshot();
  });

  describe('push should add only to collections matching filterArgumentKeys', () => {
    const initializingSchema = new schema.Collection([Todo]);
    let state = {
      ...initialState,
      ...normalize(
        initializingSchema,
        [{ id: '10', title: 'create new items' }],
        [{ userId: '1' }],
        initialState,
      ),
    };
    state = {
      ...state,
      ...normalize(
        initializingSchema,
        [{ id: '10', title: 'create new items' }],
        [{ userId: '1', ignoredMe: '5' }],
        state,
      ),
    };
    state = {
      ...state,
      ...normalize(
        initializingSchema,
        [{ id: '20', title: 'second user' }],
        [{ userId: '2' }],
        state,
      ),
    };
    state = {
      ...state,
      ...normalize(
        initializingSchema,
        [
          { id: '10', title: 'create new items' },
          { id: '20', title: 'the ignored one' },
        ],
        [{}],
        state,
      ),
    };
    function validate(sch: schema.Collection<(typeof Todo)[]>) {
      expect(
        (
          denormalize(
            sch,
            JSON.stringify({ userId: '1' }),
            state.entities,
          ) as any
        )?.length,
      ).toBe(1);
      const testState = {
        ...state,
        ...normalize(
          sch.push,
          [{ id: '30', title: 'pushed to the end' }],
          [{ userId: '1' }],
          state,
        ),
      };
      function getResponse(...args: any) {
        return denormalize(
          sch,
          sch.pk(undefined, undefined, '', args),
          testState.entities,
        ) as any;
      }
      const userOne = getResponse({ userId: '1' });
      if (!userOne || typeof userOne === 'symbol')
        throw new Error('should have a value');
      expect(userOne.length).toBe(2);
      expect(userOne[1].title).toBe('pushed to the end');

      expect(getResponse({}).length).toBe(3);
      expect(getResponse({ ignoredMe: '5', userId: '1' })?.length).toBe(2);
      expect(getResponse({ userId: '2' })?.length).toBe(1);
    }

    it('should work with function form', () => {
      const sch = new schema.Collection([Todo], {
        nonFilterArgumentKeys(key) {
          return key.startsWith('ignored');
        },
      });
      validate(sch);
    });
    it('should work with RegExp form', () => {
      const sch = new schema.Collection([Todo], {
        nonFilterArgumentKeys: /ignored/,
      });
      validate(sch);
    });
    it('should work with RegExp form', () => {
      const sch = new schema.Collection([Todo], {
        nonFilterArgumentKeys: ['ignoredMe'],
      });
      validate(sch);
    });
    it('should work with full createCollectionFilter form', () => {
      const sch = new schema.Collection([Todo], {
        createCollectionFilter:
          (...args) =>
          collectionKey =>
            Object.entries(collectionKey).every(
              ([key, value]) =>
                key.startsWith('ignored') ||
                // strings are canonical form. See pk() above for value transformation
                `${args[0]?.[key]}` === value ||
                `${args[1]?.[key]}` === value,
            ),
      });
      validate(sch);
    });
    it('should work with function override of nonFilterArgumentKeys', () => {
      class MyCollection<
        S extends any[] | PolymorphicInterface = any,
        Parent extends any[] =
          | []
          | [urlParams: { [k: string]: any }]
          | [urlParams: { [k: string]: any }, body: { [k: string]: any }],
      > extends schema.Collection<S, Parent> {
        nonFilterArgumentKeys(key: string) {
          return key.startsWith('ignored');
        }
      }
      const sch = new MyCollection([Todo]);
      validate(sch);
    });
    it('should work with function override of createCollectionFilter', () => {
      class MyCollection<
        S extends any[] | PolymorphicInterface = any,
        Parent extends any[] =
          | []
          | [urlParams: { [k: string]: any }]
          | [urlParams: { [k: string]: any }, body: { [k: string]: any }],
      > extends schema.Collection<S, Parent> {
        createCollectionFilter(...args: Parent) {
          return (collectionKey: { [k: string]: string }) =>
            Object.entries(collectionKey).every(
              ([key, value]) =>
                key.startsWith('ignored') ||
                // strings are canonical form. See pk() above for value transformation
                `${args[0][key]}` === value ||
                `${args[1]?.[key]}` === value,
            );
        }
      }
      const sch = new MyCollection([Todo]);
      validate(sch);
    });
  });
});

describe(`${schema.Collection.name} denormalization`, () => {
  const normalizeNested = {
    entities: {
      [userTodos.key]: {
        '{"userId":"1"}': ['5'],
      },
      Todo: {
        '5': {
          id: '5',
          title: 'finish collections',
        },
      },
      User: {
        '1': {
          id: '1',
          todos: '{"userId":"1"}',
          username: 'bob',
        },
      },
    },
    entityMeta: {
      [userTodos.key]: {
        '{"userId":"1"}': {
          date: 1557831718135,
          expiresAt: Infinity,
          fetchedAt: 0,
        },
      },
      Todo: {
        '5': {
          date: 1557831718135,
          expiresAt: Infinity,
          fetchedAt: 0,
        },
      },
      User: {
        '1': {
          date: 1557831718135,
          expiresAt: Infinity,
          fetchedAt: 0,
        },
      },
    },
    indexes: {},
    result: '1',
  };

  test('denormalizes nested collections', () => {
    expect(
      denormalize(User, normalizeNested.result, normalizeNested.entities),
    ).toMatchSnapshot();
  });

  test('denormalizes top level collections', () => {
    expect(
      denormalize(userTodos, '{"userId":"1"}', normalizeNested.entities),
    ).toMatchSnapshot();
  });

  describe('caching', () => {
    const memo = new SimpleMemoCache();
    test('denormalizes nested and top level share referential equality', () => {
      const todos = memo.denormalize(
        userTodos,
        '{"userId":"1"}',
        normalizeNested.entities,

        [{ userId: '1' }],
      );
      const user = memo.denormalize(
        User,
        normalizeNested.result,
        normalizeNested.entities,
      );
      expect(user).toBeDefined();
      expect(user).not.toEqual(expect.any(Symbol));
      if (typeof user === 'symbol' || !user) return;
      expect(todos).toBe(user.todos);
    });

    test('push updates cache', () => {
      const pushedState = normalize(
        User.schema.todos.push,
        [{ id: '10', title: 'create new items' }],
        [{ userId: '1' }],
        normalizeNested,
      );
      const todos = memo.denormalize(
        userTodos,
        '{"userId":"1"}',
        pushedState.entities,

        [{ userId: '1' }],
      );
      const user = memo.denormalize(
        User,
        normalizeNested.result,
        pushedState.entities,
        [{ id: '1' }],
      );
      expect(user).toBeDefined();
      expect(user).not.toEqual(expect.any(Symbol));
      if (typeof user === 'symbol' || !user) return;
      expect(user.todos.length).toBe(2);
      expect(todos).toBe(user.todos);
    });

    test('unshift places at start', () => {
      const unshiftState = normalize(
        User.schema.todos.unshift,
        [{ id: '2', title: 'from the start' }],
        [{ userId: '1' }],
        normalizeNested,
      );
      const todos = memo.denormalize(
        userTodos,
        '{"userId":"1"}',
        unshiftState.entities,
        [{ userId: '1' }],
      );
      const user = memo.denormalize(
        User,
        normalizeNested.result,
        unshiftState.entities,
        [{ id: '1' }],
      );
      expect(user).toBeDefined();
      expect(user).not.toEqual(expect.any(Symbol));
      if (typeof user === 'symbol' || !user) return;
      expect(user.todos.length).toBe(2);
      expect(todos).toBe(user.todos);
      expect(user.todos[0].title).toBe('from the start');
    });

    test('denormalizes unshift', () => {
      const todos = memo.denormalize(
        userTodos.unshift,
        [{ id: '2', title: 'from the start' }],
        {},
        [{ id: '1' }],
      );
      expect(todos).toBeDefined();
      expect(todos).not.toEqual(expect.any(Symbol));
      if (typeof todos === 'symbol' || !todos) return;
      //expect(todos[0].title).toBe('from the start'); TODO: once we get types based on parameters sent
      expect(todos).toMatchInlineSnapshot(`
        [
          Todo {
            "completed": false,
            "id": "2",
            "title": "from the start",
            "userId": 0,
          },
        ]
      `);
    });

    test('denormalizes unshift (single item)', () => {
      const todos = memo.denormalize(
        userTodos.unshift,
        { id: '2', title: 'from the start' },
        {},
        [{ id: '1' }],
      );
      expect(todos).toBeDefined();
      expect(todos).not.toEqual(expect.any(Symbol));
      if (typeof todos === 'symbol' || !todos) return;
      expect(todos.title).toBe('from the start');
      expect(todos).toMatchInlineSnapshot(`
        Todo {
          "completed": false,
          "id": "2",
          "title": "from the start",
          "userId": 0,
        }
      `);
    });
  });

  it('should buildQueryKey with matching args', () => {
    const memo = new MemoCache();
    const queryKey = memo.buildQueryKey(
      userTodos,
      [{ userId: '1' }],
      normalizeNested.entities,
      {},
    );
    expect(queryKey).toBeDefined();
    // now ensure our queryKey is usable
    const results = denormalize(userTodos, queryKey, normalizeNested.entities);
    expect(results).toBeDefined();
    expect(results).toMatchInlineSnapshot(`
      [
        Todo {
          "completed": false,
          "id": "5",
          "title": "finish collections",
          "userId": 0,
        },
      ]
    `);
  });

  it('should buildQueryKey undefined when not in cache', () => {
    const memo = new MemoCache();
    const queryKey = memo.buildQueryKey(
      userTodos,
      [{ userId: '100' }],
      normalizeNested.entities,
      {},
    );
    expect(queryKey).toBeUndefined();
  });

  it('should buildQueryKey undefined with nested Collection', () => {
    const memo = new MemoCache();
    const queryKey = memo.buildQueryKey(
      User.schema.todos,
      [{ userId: '1' }],
      normalizeNested.entities,
      {},
    );
    expect(queryKey).toBeUndefined();
  });

  it('pk should serialize differently with nested args', () => {
    const filtersA = {
      search: {
        type: 'Coupon',
      },
    };
    const filtersB = {
      search: {
        type: 'Cashback',
      },
    };

    expect(
      ArticleResource.getList.schema.pk([], undefined, '', [filtersA]),
    ).not.toEqual(
      ArticleResource.getList.schema.pk([], undefined, '', [filtersB]),
    );
  });
});
