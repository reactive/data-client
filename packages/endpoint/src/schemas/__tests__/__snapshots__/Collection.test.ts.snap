// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CollectionSchema denormalization denormalizes nested collections 1`] = `
User {
  "email": "",
  "id": "1",
  "name": "",
  "todos": [
    Todo {
      "completed": false,
      "id": "5",
      "title": "finish collections",
      "userId": 0,
    },
  ],
  "username": "bob",
}
`;

exports[`CollectionSchema denormalization denormalizes top level collections 1`] = `
[
  Todo {
    "completed": false,
    "id": "5",
    "title": "finish collections",
    "userId": 0,
  },
]
`;

exports[`CollectionSchema normalization normalizes already processed entities 1`] = `
{
  "entities": {
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
        "6",
      ],
    },
  },
  "entitiesMeta": {
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "1",
}
`;

exports[`CollectionSchema normalization normalizes nested collections 1`] = `
{
  "entities": {
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "1",
}
`;

exports[`CollectionSchema normalization normalizes push onto the end 1`] = `
{
  "entities": {
    "Todo": {
      "10": {
        "id": "10",
        "title": "create new items",
      },
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
        "10",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "10": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": [
    "10",
  ],
}
`;

exports[`CollectionSchema normalization normalizes remove from collection 1`] = `
{
  "entities": {
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
      "6": {
        "id": "6",
        "title": "another todo",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "6",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "6": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": [
    "5",
  ],
}
`;

exports[`CollectionSchema normalization normalizes remove from collection with multiple items 1`] = `
{
  "entities": {
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
      "6": {
        "id": "6",
        "title": "another todo",
      },
      "7": {
        "id": "7",
        "title": "third todo",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
        "7",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "6": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "7": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": [
    "6",
  ],
}
`;

exports[`CollectionSchema normalization normalizes remove non-existent item from collection 1`] = `
{
  "entities": {
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
      "99": {
        "id": "99",
        "title": "non-existent todo",
      },
    },
    "User": {
      "1": {
        "id": "1",
        "todos": "{"userId":"1"}",
        "username": "bob",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
      "99": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "User": {
      "1": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": [
    "99",
  ],
}
`;

exports[`CollectionSchema normalization normalizes top level collections (no args) 1`] = `
{
  "entities": {
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "{"userId":"1"}",
}
`;

exports[`CollectionSchema normalization normalizes top level collections 1`] = `
{
  "entities": {
    "Todo": {
      "5": {
        "id": "5",
        "title": "finish collections",
      },
    },
    "[Todo]": {
      "{"userId":"1"}": [
        "5",
      ],
    },
  },
  "entitiesMeta": {
    "Todo": {
      "5": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
    "[Todo]": {
      "{"userId":"1"}": {
        "date": 1557831718135,
        "expiresAt": Infinity,
        "fetchedAt": 0,
      },
    },
  },
  "indexes": {},
  "result": "{"userId":"1"}",
}
`;

exports[`CollectionSchema normalization polymorphism works with Union members 1`] = `"{"fakeFilter":"false"}"`;

exports[`CollectionSchema normalization polymorphism works with Union members 2`] = `
{
  "Group": {
    "2": {
      "id": "2",
      "type": "groups",
    },
  },
  "User": {
    "1": {
      "id": "1",
      "type": "users",
    },
  },
  "[User;Group]": {
    "{"fakeFilter":"false"}": [
      {
        "id": "1",
        "schema": "users",
      },
      {
        "id": "2",
        "schema": "groups",
      },
    ],
  },
}
`;

exports[`CollectionSchema normalization polymorphism works with polymorphic members 1`] = `"{"fakeFilter":"false"}"`;

exports[`CollectionSchema normalization polymorphism works with polymorphic members 2`] = `
{
  "Group": {
    "2": {
      "id": "2",
      "type": "groups",
    },
  },
  "User": {
    "1": {
      "id": "1",
      "type": "users",
    },
  },
  "[User;Group]": {
    "{"fakeFilter":"false"}": [
      {
        "id": "1",
        "schema": "users",
      },
      {
        "id": "2",
        "schema": "groups",
      },
    ],
  },
}
`;

exports[`CollectionSchema normalization should throw a custom error if data loads with args of unexpected value 1`] = `"Cannot destructure property 'userId' of 'undefined' as it is undefined."`;

exports[`CollectionSchema normalization should throw a custom error if data loads with string unexpected value 1`] = `
"Unexpected input given to normalize. Expected type to be "object", found "string".

          Schema: {
  "key": "[Todo]",
  "schema": {
    "key": "Todo",
    "schema": {}
  }
}
          Input: "abc""
`;

exports[`CollectionSchema normalization should throw a custom error if data loads with string unexpected value 2`] = `
"Unexpected input given to normalize. Expected type to be "object", found "null".

          Schema: {
  "key": "[Todo]",
  "schema": {
    "key": "Todo",
    "schema": {}
  }
}
          Input: "null""
`;
